src/components/datepicker/datePicker.js

```diff
@@ -0,0 +1,361 @@
+
+    this.calendarPane.id = 'md-date-pane' + $mdUtil.nextUid();
+
+    $mdTheming($element);
+
+    /** Pre-bound click handler is saved so that the event listener can be removed. */
+    this.bodyClickHandler = angular.bind(this, this.handleBodyClick);
+
+    // Unless the user specifies so, the datepicker should not be a tab stop.
+    // This is necessary because ngAria might add a tabindex to anything with an ng-model
+    // (based on whether or not the user has turned that particular feature on/off).
+    if (!$attrs['tabindex']) {
+      $element.attr('tabindex', '-1');
+    }
+
+    this.installPropertyInterceptors();
+    this.attachChangeListeners();
+    this.attachInterationListeners();
+
+    var self = this;
+    $scope.$on('$destroy', function() {
+      self.detachCalendarPane();
+    });
+  }
+
+  /**
+   * Sets up the controller's reference to ngModelController.
+   * @param {!angular.NgModelController} ngModelCtrl
+   */
+  DatePickerCtrl.prototype.configureNgModel = function(ngModelCtrl) {
+    this.ngModelCtrl = ngModelCtrl;
+
+    var self = this;
+    ngModelCtrl.$render = function() {
+      self.date = self.ngModelCtrl.$viewValue;
+      self.inputElement.value = self.dateLocale.formatDate(self.date);
+      self.inputElement.size = self.inputElement.value.length + EXTRA_INPUT_SIZE;
+    };
+  };
+
+  /**
+   * Attach event listeners for both the text input and the md-calendar.
+   * Events are used instead of ng-model so that updates don't infinitely update the other
+   * on a change. This should also be more performant than using a $watch.
+   */
+  DatePickerCtrl.prototype.attachChangeListeners = function() {
+    var self = this;
+
+    self.$scope.$on('md-calendar-change', function(event, date) {
+      self.ngModelCtrl.$setViewValue(date);
+      self.inputElement.value = self.dateLocale.formatDate(date);
+      self.closeCalendarPane();
+    });
+
+    // TODO(jelbourn): Debounce this input event.
+    self.inputElement.addEventListener('input', function() {
+      var inputString = self.inputElement.value;
+      var parsedDate = self.dateLocale.parseDate(inputString);
+
+      self.inputElement.size = inputString.length + EXTRA_INPUT_SIZE;
+
+      if (self.dateUtil.isValidDate(parsedDate)) {
+        // TODO(jelbourn): if we can detect here that `inputString` is a "complete" date,
+        // set the ng-model value.
+        self.date = parsedDate;
+        self.$scope.$apply();
+      }
+    });
+  };
+
+  /** Attach event listeners for user interaction. */
+  DatePickerCtrl.prototype.attachInterationListeners = function() {
+    var self = this;
+    var $scope = this.$scope;
+    var keyCodes = this.$mdConstant.KEY_CODE;
+
+    // Add event listener through angular so that we can triggerHandler in unit tests.
+    angular.element(self.inputElement).on('keydown', function(event) {
+      $scope.$apply(function() {
+        if (event.altKey && event.keyCode == keyCodes.DOWN_ARROW) {
```

Comment:
Should `$apply` live inside the if statement?  Or do we want to run it regardless?